presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java

```diff
@@ -0,0 +1,545 @@
+
+public class PageFunctionCompiler
+{
+    private final Metadata metadata;
+    private final DeterminismEvaluator determinismEvaluator;
+
+    @Inject
+    public PageFunctionCompiler(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.determinismEvaluator = new DeterminismEvaluator(metadata.getFunctionRegistry());
+    }
+
+    public Supplier<PageProjection> compileProjection(RowExpression projection)
+    {
+        requireNonNull(projection, "projection is null");
+
+        if (projection instanceof InputReferenceExpression) {
+            InputReferenceExpression input = (InputReferenceExpression) projection;
+            InputPageProjection projectionFunction = new InputPageProjection(input.getField(), input.getType());
+            return () -> projectionFunction;
+        }
+
+        if (projection instanceof ConstantExpression) {
+            ConstantExpression constant = (ConstantExpression) projection;
+            ConstantPageProjection projectionFunction = new ConstantPageProjection(constant.getValue(), constant.getType());
+            return () -> projectionFunction;
+        }
+
+        PageFieldsToInputParametersVisitor visitor = new PageFieldsToInputParametersVisitor();
+        RowExpression rewrittenProjection = projection.accept(visitor, null);
+        InputChannels inputChannels = new InputChannels(visitor.getInputChannels());
+
+        CallSiteBinder callSiteBinder = new CallSiteBinder();
+        ClassDefinition classDefinition = defileProjectionClass(rewrittenProjection, inputChannels, callSiteBinder);
+
+        Class<? extends PageProjection> projectionClass;
+        try {
+            projectionClass = defineClass(classDefinition, PageProjection.class, callSiteBinder.getBindings(), getClass().getClassLoader());
+        }
+        catch (Exception e) {
+            throw new PrestoException(COMPILER_ERROR, e.getCause());
+        }
+
+        return () -> {
+            try {
+                return projectionClass.newInstance();
+            }
+            catch (ReflectiveOperationException e) {
+                throw new PrestoException(COMPILER_ERROR, e);
+            }
+        };
+    }
+
+    private ClassDefinition defileProjectionClass(RowExpression projection, InputChannels inputChannels, CallSiteBinder callSiteBinder)
+    {
+        ClassDefinition classDefinition = new ClassDefinition(
+                a(PUBLIC, FINAL),
+                makeClassName(PageProjection.class.getSimpleName()),
+                type(Object.class),
+                type(PageProjection.class));
+
+        FieldDefinition blockBuilderField = classDefinition.declareField(a(PRIVATE, FINAL), "blockBuilder", BlockBuilder.class);
+
+        CachedInstanceBinder cachedInstanceBinder = new CachedInstanceBinder(classDefinition, callSiteBinder);
+        generateProjectMethod(classDefinition, callSiteBinder, cachedInstanceBinder, projection, blockBuilderField);
```

Comment:
make order of calls to 
`generateProjectMethod` and `generatePageProjectMethod` the same as order of declarations.